### PR TITLE
MSCA-771: Add winsorization option to metrics

### PR DESCRIPTION
**Pull request by kcercy**
Adds an opt-in `winsorize=(lower, upper)` parameter to `Metric.eval` and
`eval_skill`. Per-observation error contributions are clipped to the given
quantiles before aggregating, capping the influence of blowup holdouts on
skill scores. Applies to all metrics; generalizes the winsorized RMSE from
[OneMod v0.2.2](https://github.com/ihmeuw-msca/OneMod/blob/v0.2.2/src/onemod/modeling/metric.py#L60).

```python
metric.eval(df, "obs", "pred", "weights", winsorize=(0.0, 0.95))
```

- Default None leaves the existing path untouched (backward compatible).
- Grouped quantiles are computed per group; aggregation honors weights.
- Verified to match the OneMod reference under uniform weights (single and grouped). Reproducing OneMod's holdout-averaged scores additionally requires winsorizing per holdout slice and averaging afterward, rather than grouping by a coarser key.

### CODE REVIEW

File: src/msca/metrics/main.py [hunk 1]
```diff
@@ -205,10 +244,8 @@ def _eval_single(
                     sample_weight=weight_values,
                 )
             case Metric.MEDIAN_ABSOLUTE_ERROR:
-                result = metrics.median_absolute_error(
```
Comment (kcercy) [inline src/msca/metrics/main.py:208]: **For discussion: weighted median for median absolute error**

sklearn has no public weighted-median function — it only exposes one inside
`median_absolute_error(y_true, y_pred, sample_weight)`. To winsorize the median
we'd reused it via a zeros array (`median_absolute_error(contributions, 0)`),
which misuses an error metric as a median primitive and hides a "must be >= 0"
precondition.

Replaced with a small `_weighted_median(values, weights)` helper, used by
**both** the standard and winsorized median paths so they share one definition.

**Behavior:** equal-weight results unchanged; non-uniform weights now use our
convention instead of sklearn's internal one (~1e-3 difference).

**Question:** own the weighted-median definition (consistent, no private-sklearn
dependency), or keep the standard path on sklearn and accept a small
standard-vs-winsorized gap on weighted data?

Comment (zhengp0) [reply src/msca/metrics/main.py:208]: Hi @kcercy! Thanks very much for implementing this! If I read the code right, `_weighted_median` is slightly different than `sklearn` implementation https://github.com/scikit-learn/scikit-learn/blob/1.8.0/sklearn/utils/stats.py#L10-L213. I think `sklearn` try to stick with the original values. To be honest I like your implementation a little more! It smoothly interpolate the cdf, and the behavior is consistent with `np.median`. And I am ok to not working with `metrics.median_absolute_error`. Since we already implement lots of other `sklearn` metrics function due to the winsorizing, maybe later for a separate PR it is also ok to remove the dependency on sklearn, IMO.

File: src/msca/metrics/main.py
Comment (kcercy) [inline src/msca/metrics/main.py:36]: Test with data that only has few rows!

File: pyproject.toml
Comment (kcercy) [inline pyproject.toml:21]: np.quantile (diff between 1.* and 2.*), might accept weights, could use this perhaps, check versions. then we wouldn't need to own implementation.

**Review by ADucellierIHME — APPROVED** [review]
